cmd/cmd.go

```diff
@@ -1398,6 +1409,8 @@ func NewCLI() *cobra.Command {
 				envVars["OLLAMA_SCHED_SPREAD"],
 				envVars["OLLAMA_TMPDIR"],
 				envVars["OLLAMA_FLASH_ATTENTION"],
+				envVars["OLLAMA_CACHE_TYPE_K"],
```

Comment:
Does it make sense to quantize these differently? vs one `OLLAMA_CACHE_TYPE`

Reply:
As below - It's worth having sperate options for them as there's slightly different trade offs and depending on your use case you might want to use the same for both or just one of them.

Reply:
How big is the tradeoff/difference? What's the case where you'd want to quantize them differently?

Reply:
Not to revive this and drag it on more, but do you have a link as to how the differences compare? I.e. use cases where you'd want to quantize them differently?

My preference would be a single variable such as:

```
OLLAMA_KV_CACHE_TYPE=q4_0
```

And then from there we can experiment with individual k/v caches. 

The challenge with adding too many options here: there are many issues open for Ollama to support other tensor libraries optimized for specific hardware – e.g. DirectML on Windows, MLX on macOS, TRT-LLM on NVIDIA hardware.

These backends do not or may not support individual separate K, V quantization, making it hard to enable this feature on all platforms in the future (and it's a good feature, so we'd want to avoid edge cases where it's available on one but not the other).

Reply:
I've seen a few folks mention using higher precision for keys and lower for values to balance the tradeoff.

These are my notes on it:

## K vs V cache quantisation

Precision Requirements:

- Keys are used in attention calculations to determine which parts of the input are relevant. This operation is more sensitive to precision loss.
- Values store the actual content to be retrieved. They can often tolerate more aggressive quantisation without significantly impacting model quality.

Efficiency:

- Key-query dot products in attention mechanisms benefit from higher precision to maintain accuracy in similarity calculations.
- Values are typically used in linear combinations, which are more robust to lower precision.

Memory Bandwidth Optimisation:

- Keys are accessed more frequently during inference (for every token generated), so keeping them in higher precision can reduce the need for frequent dequantisation.
- Values are only accessed when their corresponding keys are selected, allowing for more aggressive compression to save memory bandwidth.

Architecture Considerations:

- Some architectures (like MQA (multi-query attention)) use fewer keys than values, making it more feasible to keep keys in higher precision.
- The relative importance of keys vs. values can vary depending on the specific attention mechanism used.

Reply:
As far as benchmarked differences between them - I don't know that I've seen numbers put to paper as such.

Reply:
Also see - https://github.com/ollama/ollama/pull/6279/files/0864c874684f11cd2291114457814a5d15fb250f#r1741205838

Reply:
After considering and seeing similar LLM serving tooling, we should make this one datatype for now (vs separate k/v types). I'm not against adding separate K & V quantization in the future, but as we add more potential hardware support, we'd want to make sure the kernels are easier to support on respective platforms – having to deal with combinations between different k/v quantization would make this harder

Reply:
OK I'll refactor it all again.